### ISSUE TITLE
PowerPC Power8 opencv capture video Color deviation

### ISSUE THREAD

**Issue by HaitaoYue** [opencv/opencv#11034]
Architecture: ppc64le
OS: Linux (ubuntu)
Compiler: GCC
Run environment: docker
Docker version: 1.12.6
Opencv Version: 3.3.0
ffmpeg version: 3.3.4
Dockerfile: https://gist.github.com/HaitaoYue/41370517b54555ab5e9abfdbee8f38d7
python test file: https://gist.github.com/HaitaoYue/a9c115ad612af29d7b4207e9004c6bdb

Result image compare:
Right result:
![2018-03-08 9 50 52](https://user-images.githubusercontent.com/35512917/37128966-a082f780-22b7-11e8-8f2b-c0d307e7453c.png)
Wrong result:
![wrong](https://user-images.githubusercontent.com/35512917/37129178-99f910a6-22b8-11e8-8786-3f7d519fd80b.jpg)

**Comment by alalek** [comment]
Did "ffplay" or frames extraction via "ffmpeg" work well on your configuration (without OpenCV)?

**Comment by HaitaoYue** [comment]
@alalek yes, frames extraction via "ffmpeg" is fine.

**Comment by seiko2plus** [comment]
> yes, frames extraction via "ffmpeg" is fine.

@HaitaoYue, Are sure? could try to extract the frame with alpha support? like PNG?

I found a bug on libswscale (colorspace conversion optimization) , to work around it you have to build ffmpeg without altivec support "--disable-altivec  --disable-vsx" and this problem should gone.

@alalek, videoio/Videoio_Synthetic test fails when link against libswscale with altivec/vsx support.

**Comment by HaitaoYue** [comment]
@seiko2plus I have tried to save to png and compile the ffmpeg with "--disable-altivec --disable-vsx" but it's not worked, same as before.

**Comment by seiko2plus** [comment]
@HaitaoYue, I think this issue should forward to ffmpeg/libav since OpenCV isn't involved at all.
It seems that PPC/Altivec SIMD code in libswscale missing little-endian support.

-> I have tried to save to png and compile the ffmpeg with "--disable-altivec --disable-vsx" but it's not worked, same as before.
That's too weird, Please Could you provide me a docker file contains your testing process?

Anyway here's my testing result, [Docker file](https://raw.githubusercontent.com/seiko2plus/just4share/master/opencv/issue11034/Dockerfile)


#### Environment:

````
cat /proc/cpuinfo

processor	: 0
cpu		: POWER8 (architected), altivec supported
clock		: 3425.000000MHz
revision	: 2.1 (pvr 004b 0201)

processor	: 1
cpu		: POWER8 (architected), altivec supported
clock		: 3425.000000MHz
revision	: 2.1 (pvr 004b 0201)

timebase	: 512000000
platform	: pSeries
model		: IBM pSeries (emulated by qemu)
machine		: CHRP IBM pSeries (emulated by qemu)
MMU		: Hash
````

````
uname -a
Linux 9a1f29a6f3f5 4.14.0-3-powerpc64le #1 SMP Debian 4.14.13-1 (2018-01-14) ppc64le GNU/Linu
````

````
cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux 9 (stretch)"
NAME="Debian GNU/Linux"
VERSION_ID="9"
VERSION="9 (stretch)"
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
````

````
ffmpeg -version
ffmpeg version n3.4.2 Copyright (c) 2000-2018 the FFmpeg developers
built with gcc 6.3.0 (Debian 6.3.0-18+deb9u1) 20170516
configuration: --prefix=/usr --extra-libs='-lpthread -lm' --enable-gpl --enable-libvpx --enable-libx264 --enable-nonfree --enable-shared --enable-avresample --enable-debug=3 --disable-doc --disable-static
libavutil      55. 78.100 / 55. 78.100
libavcodec     57.107.100 / 57.107.100
libavformat    57. 83.100 / 57. 83.100
libavdevice    57. 10.100 / 57. 10.100
libavfilter     6.107.100 /  6.107.100
libavresample   3.  7.  0 /  3.  7.  0
libswscale      4.  8.100 /  4.  8.100
libswresample   2.  9.100 /  2.  9.100
libpostproc    54.  7.100 / 54.  7.100
````

````
opencv_version
3.4.1
````

**Comment by seiko2plus** [comment]
#### Testing result before disable optimization:

##### Using ffmpeg command, [Video URL](http://commondatastorage.googleapis.com/gtv-videos-bucket/sample/Sintel.mp4)

````
ffmpeg -i ./Sintel.mp4 -filter:v fps=fps=1/60 /tests/before/frames/%0d.png
````


###### PNG Simple
![8](https://user-images.githubusercontent.com/12713707/37908147-91d20182-3107-11e8-9776-381fc03e1516.png)

````
ffmpeg -i ./Sintel.mp4 -filter:v fps=fps=1/60 /tests/before/frames/%0d.jpg
````

###### JPEG Simple
![8](https://user-images.githubusercontent.com/12713707/37908177-b38b0b52-3107-11e8-905b-f465f15ccef2.jpg)



##### Using OpenCV, [Source code URL](https://raw.githubusercontent.com/seiko2plus/just4share/master/opencv/issue11034/cv_grep_frame.cpp)

cv_grep_frame ./Sintel.mp4 -output=/tests/before/cv_frames/ -fes=60 -pos=29

###### PNG Simple
![8](https://user-images.githubusercontent.com/12713707/37908214-d2473b88-3107-11e8-9cce-1a39fe759d62.png)

###### JPEG Simple
![8](https://user-images.githubusercontent.com/12713707/37908224-da863d26-3107-11e8-88cd-4b0b1302f3fc.jpg)

#####


##### ffmpeg tests (make check) doesn't report any errors

````
tail -n 10 /tests/before/ffmpeg_check.log
TEST    utvideoenc_yuv422_none
COPY	tests/test_copy.ffmeta
GEN	tests/data/ffprobe-test.nut
TEST    ffprobe_compact
TEST    ffprobe_csv
TEST    ffprobe_default
TEST    ffprobe_flat
TEST    ffprobe_ini
TEST    ffprobe_json
TEST    ffprobe_xml
````

##### opencv_videoio test summary

````
tail -n 20 /tests/before/opencv_test_videoio.log
[----------] Global test environment tear-down
[==========] 47 tests from 6 test cases ran. (134420 ms total)
[  PASSED  ] 33 tests.
[  FAILED  ] 14 tests, listed below:
[  FAILED  ] videoio/Videoio_Synthetic.write_read_position/0, where GetParam() = (640x480, ("avi", "XVID", 30, 1900))
[  FAILED  ] videoio/Videoio_Synthetic.write_read_position/1, where GetParam() = (640x480, ("avi", "MPEG", 30, 1900))
[  FAILED  ] videoio/Videoio_Synthetic.write_read_position/2, where GetParam() = (640x480, ("avi", "IYUV", 30, 1900))
[  FAILED  ] videoio/Videoio_Synthetic.write_read_position/3, where GetParam() = (640x480, ("avi", "MJPG", 30, 1900))
[  FAILED  ] videoio/Videoio_Synthetic.write_read_position/4, where GetParam() = (640x480, ("mkv", "XVID", 30, 1900))
[  FAILED  ] videoio/Videoio_Synthetic.write_read_position/5, where GetParam() = (640x480, ("mkv", "MPEG", 30, 1900))
[  FAILED  ] videoio/Videoio_Synthetic.write_read_position/6, where GetParam() = (640x480, ("mkv", "MJPG", 30, 1900))
[  FAILED  ] videoio/Videoio_Synthetic.write_read_position/8, where GetParam() = (976x768, ("avi", "XVID", 30, 1900))
[  FAILED  ] videoio/Videoio_Synthetic.write_read_position/9, where GetParam() = (976x768, ("avi", "MPEG", 30, 1900))
[  FAILED  ] videoio/Videoio_Synthetic.write_read_position/10, where GetParam() = (976x768, ("avi", "IYUV", 30, 1900))
[  FAILED  ] videoio/Videoio_Synthetic.write_read_position/11, where GetParam() = (976x768, ("avi", "MJPG", 30, 1900))
[  FAILED  ] videoio/Videoio_Synthetic.write_read_position/12, where GetParam() = (976x768, ("mkv", "XVID", 30, 1900))
[  FAILED  ] videoio/Videoio_Synthetic.write_read_position/13, where GetParam() = (976x768, ("mkv", "MPEG", 30, 1900))
[  FAILED  ] videoio/Videoio_Synthetic.write_read_position/14, where GetParam() = (976x768, ("mkv", "MJPG", 30, 1900))

14 FAILED TESTS
````

**Comment by seiko2plus** [comment]
#### Testing result after disable optimization:

I tested with disable-altivec, also by disable yuv2rgb SIMD code using [this patch](https://raw.githubusercontent.com/seiko2plus/just4share/master/opencv/issue11034/libswscale-disable-ppc-yuv2rgb.diff) and both show same results.

##### Using ffmpeg command, [Video URL](http://commondatastorage.googleapis.com/gtv-videos-bucket/sample/Sintel.mp4)

````
ffmpeg -i ./Sintel.mp4 -filter:v fps=fps=1/60 /tests/after/frames/%0d.png
````


###### PNG Simple
![8](https://user-images.githubusercontent.com/12713707/37908811-89b84e5a-3109-11e8-8806-cc180467b204.png)

````
ffmpeg -i ./Sintel.mp4 -filter:v fps=fps=1/60 /tests/after/frames/%0d.jpg
````

###### JPEG Simple
![8](https://user-images.githubusercontent.com/12713707/37908828-9523f528-3109-11e8-877f-124207af2959.jpg)


##### Using OpenCV, [Source code URL](https://raw.githubusercontent.com/seiko2plus/just4share/master/opencv/issue11034/cv_grep_frame.cpp)

cv_grep_frame ./Sintel.mp4 -output=/tests/after/cv_frames/ -fes=60 -pos=29

###### PNG Simple
![8](https://user-images.githubusercontent.com/12713707/37908869-b0c6fad2-3109-11e8-87cc-aac3e2edf7d9.png)

###### JPEG Simple
![8](https://user-images.githubusercontent.com/12713707/37908917-cf3309ac-3109-11e8-9d33-ef6ff27a4928.jpg)

#####


##### ffmpeg tests (make check) doesn't report any errors

````
tail -n 10 /tests/after/ffmpeg_check.log
TEST    utvideoenc_yuv422_median
TEST    utvideoenc_yuv422_none
GEN	tests/data/ffprobe-test.nut
TEST    ffprobe_compact
TEST    ffprobe_csv
TEST    ffprobe_default
TEST    ffprobe_flat
TEST    ffprobe_ini
TEST    ffprobe_json
TEST    ffprobe_xml
````

##### opencv_videoio test summary

````
tail -n 20 /tests/after/opencv_test_videoio.log
[       OK ] videoio/Videoio_Synthetic.write_read_position/8 (5176 ms)
[ RUN      ] videoio/Videoio_Synthetic.write_read_position/9, where GetParam() = (976x768, ("avi", "MPEG", 30, 1900))
[       OK ] videoio/Videoio_Synthetic.write_read_position/9 (5396 ms)
[ RUN      ] videoio/Videoio_Synthetic.write_read_position/10, where GetParam() = (976x768, ("avi", "IYUV", 30, 1900))
[       OK ] videoio/Videoio_Synthetic.write_read_position/10 (4490 ms)
[ RUN      ] videoio/Videoio_Synthetic.write_read_position/11, where GetParam() = (976x768, ("avi", "MJPG", 30, 1900))
[       OK ] videoio/Videoio_Synthetic.write_read_position/11 (6828 ms)
[ RUN      ] videoio/Videoio_Synthetic.write_read_position/12, where GetParam() = (976x768, ("mkv", "XVID", 30, 1900))
[       OK ] videoio/Videoio_Synthetic.write_read_position/12 (5155 ms)
[ RUN      ] videoio/Videoio_Synthetic.write_read_position/13, where GetParam() = (976x768, ("mkv", "MPEG", 30, 1900))
[       OK ] videoio/Videoio_Synthetic.write_read_position/13 (5351 ms)
[ RUN      ] videoio/Videoio_Synthetic.write_read_position/14, where GetParam() = (976x768, ("mkv", "MJPG", 30, 1900))
[       OK ] videoio/Videoio_Synthetic.write_read_position/14 (6792 ms)
[ RUN      ] videoio/Videoio_Synthetic.write_read_position/15, where GetParam() = (976x768, ("avi", "MJPG", 30, 2200))
[       OK ] videoio/Videoio_Synthetic.write_read_position/15 (8625 ms)
[----------] 16 tests from videoio/Videoio_Synthetic (75775 ms total)

[----------] Global test environment tear-down
[==========] 47 tests from 6 test cases ran. (134368 ms total)
[  PASSED  ] 47 tests.

**Comment by HaitaoYue** [comment]
@seiko2plus great, thanks very much, I used your Dockerfile, and capture frame in python code, it's worked. But in my dockerfile I copy you compile parameter and patch the ffmpeg, it's can not work. maybe I missed or added some parameter for compile?

**Comment by seiko2plus** [comment]
@HaitaoYue, try to avoid any installation of libav distro packages and make sure opencv linking against the compiled version.
````
libavcodec-dev libavformat-dev libswscale-dev
````
I opened a new ticket at ffmpeg's issue tracker so any further discussion about this issue should move to [ffmpeg/7124](https://trac.ffmpeg.org/ticket/7124)

**Comment by ChipKerchner** [comment]
I spent some time looking at the cause of this bug. It looks like it is a bug with the gcc compiler for Power in versions 6.x and 7.x (verified with 6.3 and 7.4). It was fixed in version 8.x (verified with 8.3). I was using ffmpeg version 4.1 and a Power 9 ppc64le machine for building and testing.

It appears the compiler is generating the wrong code for little endian machines for the vec_lvsl/vec_perm instruction combos in some cases. If these instructions are replaced with vec_xl, the problem goes away for all versions of the compilers. It should be noted that the compiler is producing warnings for using vec_lvsl being deprecated for little endian.
